carball/analysis/events/carry_detection.py

```diff
@@ -0,0 +1,187 @@
+        """
+        Filters out all frames that are not close enough to the player to be a valid carry.
+        Filters out dribbles that are too short
+        """
+
+        xy_distance = ((player_frames.pos_x - valid_frames.ball.pos_x) ** 2 +
+                       (player_frames.pos_y - valid_frames.ball.pos_y) ** 2) ** 0.5
+        carry_frames = valid_frames[(xy_distance < BALL_SIZE) & (player_frames.pos_z < valid_frames.ball.pos_z)]
+
+        player_carry_data = CarryData(valid_frames, *self.creat_start_end_frames(carry_frames))
+        return player_carry_data, xy_distance, player_frames
+        # look up hits.
+        # any hits by the same player within a continous set of valid frames should count as carrys
+
+    def merge_dribbles(self, carry_data: CarryData, player_carry_data: CarryData):
+        """
+        Merges separated dribbles that are still within the valid dribble set.
+        """
+        start_frames = player_carry_data.start_frames
+        end_frames = player_carry_data.end_frames
+
+        merged_start = []
+        merged_end = []
+
+        previous_start = -1
+        previous_end = 1
+
+        for frame_index in range(len(start_frames) - 1):
+            total_frame_number = frame_index
+            while total_frame_number < len(carry_data.start_frames) and carry_data.start_frames[total_frame_number] < start_frames[frame_index]:
+                total_frame_number += 1
+            if total_frame_number == len(carry_data.start_frames):
+                logger.warning("can not merge frames anymore.")
+                break
+
+            end_frame = end_frames[frame_index]
+            next_start = start_frames[frame_index + 1]
+            if (end_frame < carry_data.start_frames[total_frame_number] < next_start or
+                    end_frame < carry_data.end_frames[total_frame_number] < next_start):
+                if previous_start != -1:
+                    merged_start.append(previous_start)
+                    merged_end.append(previous_end)
+                    previous_start = -1
+                else:
+                    merged_start.append(start_frames[frame_index])
+                    merged_end.append(end_frames[frame_index])
+            else:
+                if previous_start == -1:
+                    previous_start = start_frames[frame_index]
+                previous_end = end_frames[frame_index + 1]
+
+        if previous_start != -1:
+            merged_start.append(previous_start)
+            merged_end.append(previous_end)
+        player_carry_data.start_frames = merged_start
+        player_carry_data.end_frames = merged_end
+
+    def expand_carries(self, carry_data: CarryData, player_carry_data: CarryData,
+                       player: Player, proto_game: game_pb2.Game) -> CarryData:
+        """
+        Expands / merges carries based on hit list.
+        If nothing is modified then this returns valid_frames
```

Comment:
This docstring confuses me more than it explains stuff. What's "expands"? What's valid_frames?